internal/server/v1/propertyvalues/property_values.go

```diff
@@ -0,0 +1,287 @@
+		rawEntities:    nil,
+		mergedEntities: nil,
+		limit:          limit,
+		heap:           &entityHeap{},
+		totalPage:      map[int]int{},
+	}
+	// Read cache data based on pagination info.
+	rowListMap := map[int]cbt.RowList{}
+	for _, c := range cursorGroup.Cursors {
+		if c != nil {
+			rowList := buildSimpleRequestRowList(entity, property, false, c.GetPage())
+			rowListMap[int(c.GetIg())] = rowList
+		}
+	}
+	btDataList, err := bigtable.ReadWithGroupRowList(
+		ctx,
+		btGroup,
+		rowListMap,
+		func(jsonRaw []byte) (interface{}, error) {
+			var p pb.PagedEntities
+			err := proto.Unmarshal(jsonRaw, &p)
+			return &p, err
+		},
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+	// Store the raw cache data.
+	for idx, btData := range btDataList {
+		if data, ok := btData[entity]; ok {
+			pe := data.(*pb.PagedEntities)
+			p.rawEntities = append(p.rawEntities, pe.Entities)
+			p.totalPage[idx] = int(pe.TotalPageCount)
+		} else {
+			p.rawEntities = append(p.rawEntities, nil)
+		}
+	}
+	// Init the min heap.
+	heap.Init(p.heap)
+	// Push the next entity of each import group to the heap.
+	for idx, entityList := range p.rawEntities {
+		cursor := cursorGroup.Cursors[idx]
+		if int(cursor.GetItem()) < len(entityList) {
+			elem := &heapElem{
+				ig:   idx,
+				pos:  cursor.GetItem(),
+				data: entityList[cursor.GetItem()],
+			}
+			heap.Push(p.heap, elem)
+		}
+	}
+	return p, nil
+}
+
+// Process the next entity.
+//
+// Import groups are ordered by preferences, so all entities from the preferred
+// import groups should be added first. Additional entities from other non
+// preferred import groups are added by the preference order.
+//
+// As the data in each import group is sorted already. The duplication is
+// handled by advancing the pointer in each import group simounateously like
+// in a merge sort.
+func (p *processor) next(ctx context.Context, btGroup *bigtable.Group) (bool, error) {
+	// All entities in all import groups have been exhausted.
+	if p.heap.Len() == 0 {
+		return false, nil
+	}
+	elem := heap.Pop(p.heap).(*heapElem)
+	entity, ig := elem.data, elem.ig
+	if len(p.mergedEntities) == 0 {
+		p.mergedEntities = []*pb.EntityInfo{entity}
+	} else {
+		prev := p.mergedEntities[len(p.mergedEntities)-1]
+		if entity.Dcid != prev.Dcid || entity.Value != prev.Value {
+			// Find a new entity, add to the result.
+			p.mergedEntities = append(p.mergedEntities, entity)
+		}
+	}
+	// Got enough entities, should stop the process.
+	// Cursor is now at the next read item, ready to be returned.
+	if len(p.mergedEntities) == p.limit+1 {
+		p.mergedEntities = p.mergedEntities[:p.limit]
+		return false, nil
+	}
+	// Update the cursor.
+	cursor := p.cursorGroup.Cursors[ig]
+	cursor.Item++
+	// Reach the end of the current page, should advance to next page.
+	if int(cursor.Item) == len(p.rawEntities[ig]) {
+		cursor.Page++
+		cursor.Item = 0
+	}
+	// Need to update raw data if advancing to the next page.
```

Comment:
Should this if cursor.Item == 0 {} code block just be within the above code block? (just before that you increment cursor.Item++)